src/data_cleaning.py

```diff
@@ -142,8 +125,195 @@ def generate_subplant_ids(start_year, end_year, cems_monthly):
+        complete_generator_ids,
+        how="outer",
+        on=["plant_id_eia", "generator_id"],
+        validate="m:1",
+    )
+    # also add a complete list of cems unitid
+    subplant_crosswalk_complete = subplant_crosswalk_complete.merge(
+        cems_ids[["plant_id_eia", "unitid"]].drop_duplicates(),
+        how="outer",
+        on=["plant_id_eia", "unitid"],
+        validate="m:1",
+    )
+    # update the subplant ids for each plant
+    subplant_crosswalk_complete = subplant_crosswalk_complete.groupby(
+        "plant_id_eia"
+    ).apply(update_subplant_ids)
+
+    # remove the intermediate columns created by update_subplant_ids
+    subplant_crosswalk_complete["subplant_id"].update(
+        subplant_crosswalk_complete["new_subplant"]
+    )
+    subplant_crosswalk_complete = subplant_crosswalk_complete.reset_index(drop=True)[
+        [
+            "plant_id_epa",
+            "unitid",
+            "plant_id_eia",
+            "generator_id",
+            "subplant_id",
+            "unit_id_pudl",
+        ]
+    ]
+
+    subplant_crosswalk_complete = subplant_crosswalk_complete.drop_duplicates(
+        subset=[
+            "plant_id_epa",
+            "unitid",
+            "plant_id_eia",
+            "generator_id",
+            "subplant_id",
+        ],
+        keep="last",
+    )
+
     # add proposed operating dates and retirements to the subplant id crosswalk
-    pudl_db = "sqlite:///../data/downloads/pudl/pudl_data/sqlite/pudl.sqlite"
+    subplant_crosswalk_complete = add_operating_and_retirement_dates(
+        subplant_crosswalk_complete, start_year, end_year
+    )
+
+    os.makedirs(outputs_folder(f"{end_year}"), exist_ok=True)
+
+    # export the crosswalk to csv
+    subplant_crosswalk_complete.to_csv(
+        outputs_folder(f"{end_year}/subplant_crosswalk_{end_year}.csv"),
+        index=False,
+    )
+
+
+def update_subplant_ids(subplant_crosswalk):
+    """
+    Ensures a complete and accurate subplant_id mapping for all generators.
+
+    NOTE:
+        1. This function is a temporary placeholder until the `pudl.analysis.epa_crosswalk` code is updated.
+        2. This function is meant to be applied using a .groupby("plant_id_eia").apply() function. This function
+        will only properly work when applied to a single plant_id_eia at a time.
+
+    Data Preparation
+        Because the existing subplant_id crosswalk was only meant to map CAMD units to EIA generators, it
+        is missing a large number of subplant_ids for generators that do not report to CEMS. Before applying this
+        function to the subplant crosswalk, the crosswalk must be completed with all generators by outer
+        merging in the complete list of generators from EIA-860 (specifically the gens_eia860 table from pudl).
+        This dataframe also contains the complete list of `unit_id_pudl` mappings that will be necessary.
+
+    This function follows several steps:
```

Comment:
Could you either add a high-level summary of the hierarchy of subplant ID assignments? My understanding is that this hierarchy is: 
1) use the PUDL subplant_id. (in the case where the PUDL unit_id groups several subplants, use the PUDL unit_id) 
2) where there is no PUDL subplant_id, use the PUDL unit_id as the subplant_id
3) where there is neither a pudl subplant_id nor unit_id, use the generator ID as the subplant_id

Reply:
Nice, I like this summary! I'll add something like this.